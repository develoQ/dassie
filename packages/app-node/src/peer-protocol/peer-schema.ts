import {
  Infer,
  choice,
  ia5String,
  octetString,
  sequence,
  sequenceOf,
  uint64Bigint,
} from "@xen-ilp/lib-oer"

const HELLO_NEIGHBOR_PROOF_LENGTH = 32

export type PeerMessage = Infer<typeof peerMessage>

export const peerSignedHello = sequence({
  nodeId: ia5String(),
  sequence: uint64Bigint(),
  url: ia5String(),
  neighbors: sequenceOf(
    sequence({
      nodeId: ia5String(),
      proof: octetString(HELLO_NEIGHBOR_PROOF_LENGTH),
    })
  ),
})

export const peerSignedLinkStateUpdate = sequence({
  nodeId: ia5String(),
  sequence: uint64Bigint(),
  neighbors: sequenceOf(
    sequence({
      nodeId: ia5String(),
    })
  ),
})

export const peerMessage = choice({
  hello: sequence({
    signed: octetString().containing(peerSignedHello),
    signature: octetString(),
  }).tag(0),
  linkStateUpdate: sequence({
    signed: octetString().containing(peerSignedLinkStateUpdate),
    signature: octetString(),
  }).tag(1),
})
