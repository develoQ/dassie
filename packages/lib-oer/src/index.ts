export type { Infer, OerType, AnyOerType, ParseOptions } from "./base-type"
export * from "./bitstring"
export * from "./boolean"
export * from "./choice"
export * from "./enumerated"
export * from "./integer-as-bigint"
export * from "./integer-as-number"
export * from "./octet-string"
export * from "./sequence-of"
export * from "./sequence"
export * from "./string"
