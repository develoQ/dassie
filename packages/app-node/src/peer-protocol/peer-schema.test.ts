import { describe, test } from "vitest"

import { hexToUint8Array } from "@xen-ilp/lib-oer"

import { peerMessage } from "./peer-schema"

describe("xen message codec", () => {
  test("should parse a hello message", ({ expect }) => {
    const uint8Array = hexToUint8Array(
      `80 4a 02 6e 31 00 00 01 82 4b a9 32 a2 19 68 74 74 70 73 3a 2f 2f 6e 31 
       2e 6c 6f 63 61 6c 68 6f 73 74 3a 34 30 30 30 01 01 02 6e 32 00 00 00 00 
       00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
       00 00 00 00 40 6b f2 24 e8 36 b3 71 02 77 31 f9 f5 cc 05 83 cc 88 d7 58 
       f8 80 ea e3 5a bc c4 a9 08 b6 b0 5f a0 1c 68 b6 1a a7 ec 86 7c 14 97 ef 
       76 84 f4 9e 1e 27 05 f7 06 e5 ae 68 82 c6 48 df 41 3e aa 50 07`
    )

    const message = peerMessage.parse(uint8Array)

    expect(message).toMatchSnapshot()
  })
})
