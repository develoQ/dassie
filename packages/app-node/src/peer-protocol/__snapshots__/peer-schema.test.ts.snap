// Vitest Snapshot v1

exports[`xen message codec > should parse a hello message 1`] = `
{
  "length": 141,
  "success": true,
  "value": {
    "hello": {
      "signature": Uint8Array [
        107,
        242,
        36,
        232,
        54,
        179,
        113,
        2,
        119,
        49,
        249,
        245,
        204,
        5,
        131,
        204,
        136,
        215,
        88,
        248,
        128,
        234,
        227,
        90,
        188,
        196,
        169,
        8,
        182,
        176,
        95,
        160,
        28,
        104,
        182,
        26,
        167,
        236,
        134,
        124,
        20,
        151,
        239,
        118,
        132,
        244,
        158,
        30,
        39,
        5,
        247,
        6,
        229,
        174,
        104,
        130,
        198,
        72,
        223,
        65,
        62,
        170,
        80,
        7,
      ],
      "signed": {
        "neighbors": [
          {
            "nodeId": "n2",
            "proof": Uint8Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
          },
        ],
        "nodeId": "n1",
        "sequence": 1659126756002n,
        "url": "https://n1.localhost:4000",
      },
    },
  },
}
`;
